Supporting different forms of prepared statement query params

As of now if i give a prepared statement of the below form and i pass in the $1 argument, it works as expected

```sql
    select * 
    from (
		select 1 id
		union all
		select 2 id
		union all
		select 3 id
    )
    where id > $1
```

but if i send the below query I get the error `sql: expected 0 arguments, got 1` from the client side (notice i am using ? for params)

```sql
    select * 
    from (
		select 1 id
		union all
		select 2 id
		union all
		select 3 id
    )
    where id > ?
```

Logs from psql-wire

```
2022-12-12T11:16:45.662+0530	INFO	psql-wire@v0.4.0/wire.go:88	serving incoming connections	{"addr": "[::]:35432"}
2022-12-12T11:16:45.665+0530	DEBUG	psql-wire@v0.4.0/wire.go:126	serving a new client connection
2022-12-12T11:16:45.666+0530	DEBUG	psql-wire@v0.4.0/handshake.go:137	attempting to upgrade the client to a TLS connection
2022-12-12T11:16:45.666+0530	DEBUG	psql-wire@v0.4.0/handshake.go:140	no TLS certificates available continuing with a insecure connection
2022-12-12T11:16:45.666+0530	ERROR	psql-wire@v0.4.0/wire.go:116	an unexpected error got returned while serving a client connection	{"error": "EOF"}
github.com/jeroenrinzema/psql-wire.(*Server).Serve.func2
	/Users/kishanb/go/pkg/mod/github.com/jeroenrinzema/psql-wire@v0.4.0/wire.go:116
2022-12-12T11:16:45.666+0530	DEBUG	psql-wire@v0.4.0/wire.go:126	serving a new client connection
2022-12-12T11:16:45.666+0530	DEBUG	psql-wire@v0.4.0/handshake.go:137	attempting to upgrade the client to a TLS connection
2022-12-12T11:16:45.667+0530	DEBUG	psql-wire@v0.4.0/handshake.go:140	no TLS certificates available continuing with a insecure connection
2022-12-12T11:16:45.667+0530	ERROR	psql-wire@v0.4.0/wire.go:116	an unexpected error got returned while serving a client connection	{"error": "EOF"}
github.com/jeroenrinzema/psql-wire.(*Server).Serve.func2
	/Users/kishanb/go/pkg/mod/github.com/jeroenrinzema/psql-wire@v0.4.0/wire.go:116
2022-12-12T11:16:45.667+0530	DEBUG	psql-wire@v0.4.0/wire.go:126	serving a new client connection
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/wire.go:137	handshake successfull, validating authentication
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:68	reading client parameters
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:86	client parameter	{"key": "user", "value": "postgres"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:86	client parameter	{"key": "database", "value": "postgres"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/auth.go:33	authenticating client connection
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/wire.go:150	connection authenticated, writing server parameters
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:102	writing server parameters
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:113	server parameter	{"key": "is_superuser", "value": "off"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:113	server parameter	{"key": "session_authorization", "value": "postgres"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:113	server parameter	{"key": "server_encoding", "value": "UTF8"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/handshake.go:113	server parameter	{"key": "client_encoding", "value": "UTF8"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/command.go:39	ready for query... starting to consume commands
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/command.go:65	incoming command	{"length": 149, "type": "P"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/command.go:270	incoming extended query	{"query": "\n      select * \n      from (\n\t\t  select 1 id\n\t\t  union all\n\t\t  select 2 id\n\t\t  union all\n\t\t  select 3 id\n      )\n      where id > ?\n   ", "name": "pgx_0", "parameters": 0}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/command.go:65	incoming command	{"length": 11, "type": "D"}
2022-12-12T11:16:45.668+0530	DEBUG	psql-wire@v0.4.0/command.go:65	incoming command	{"length": 4, "type": "S"}
    rdapp_test.go:94: 
        	Error Trace:	/Users/kishanb/Programming/Personal/rdapp/componenttest/rdapp_test.go:94
        	Error:      	Received unexpected error:
        	            	sql: expected 0 arguments, got 1
        	Test:       	TestRedshiftDataAPIProxyTestSuite/TestPreparedStatementQueryExecution
--- FAIL: TestRedshiftDataAPIProxyTestSuite (0.01s)
    --- FAIL: TestRedshiftDataAPIProxyTestSuite/TestPreparedStatementQueryExecution (0.01s)


FAIL
2022-12-12T11:16:45.669+0530	DEBUG	psql-wire@v0.4.0/command.go:65	incoming command	{"length": 23, "type": "Q"}
2022-12-12T11:16:45.669+0530	DEBUG	psql-wire@v0.4.0/command.go:216	incoming simple query	{"query": "deallocate \"pgx_0\""}
```

If you notice the log line `incoming extended query` it says `parameters` as 0

This can be simulated by using the example code from the stdlib `database/sql` package

Link to code: https://pkg.go.dev/database/sql#DB.Prepare  (screenshot below for easy access)

<img width="800" alt="image" src="https://user-images.githubusercontent.com/763760/206970423-7c87d513-0d75-4f1b-8d61-3c01cd477bcd.png">

--- a/options.go
+++ b/options.go
@@ -13,11 +13,11 @@ import (
 	"go.uber.org/zap"
 )
 
-// PositionalParameter represents the reserved character used to define a
-// positional parameter. Positional parameters are defined inside a query using
-// a dollar sign followed by a digit.
+// QueryParameters represents a regex which could be used to identify and lookup
+// parameters defined inside a given query. Parameters could be defined as
+// positional parameters and un-positional parameters.
 // https://www.postgresql.org/docs/8.1/sql-syntax.html#:~:text=A%20dollar%20sign%20(%24)%20followed,a%20dollar%2Dquoted%20string%20constant.
-var PositionalParamter = regexp.MustCompile(`\$(\d+)`)
+var QueryParameters = regexp.MustCompile(`\$(\d+)|\?`)
 
 // SimpleQueryFn represents a callback function called whenever an incoming
 // query is being executed.
@@ -68,15 +68,23 @@ func SimpleQuery(fn SimpleQueryFn) OptionFn {
 				return fn(ctx, query, writer, parameters)
 			}
 
-			// NOTE: we have to lookup all unique positional parameters
-			// within the given query. We return a zero parameter oid for each
-			// positional parameter indicating that the given parameters could
-			// contain any type. We could safely ignore the err check while
-			// converting given parameters since ony positions are returned by
-			// the positional parameter regex.
-			positions := PositionalParamter.FindAllStringSubmatch(query, -1)
-			parameters := make([]oid.Oid, 0, len(positions))
-			for _, match := range positions {
+			// NOTE: we have to lookup all parameters within the given query.
+			// Parameters could represent positional parameters or anonymous
+			// parameters. We return a zero parameter oid for each parameter
+			// indicating that the given parameters could contain any type. We
+			// could safely ignore the err check while converting given
+			// parameters since ony matches are returned by the positional
+			// parameter regex.
+			matches := QueryParameters.FindAllStringSubmatch(query, -1)
+			parameters := make([]oid.Oid, 0, len(matches))
+			for _, match := range matches {
+				// NOTE: we have to check whether the returned match is a
+				// positional parameter or an un-positional parameter.
+				// SELECT * FROM users WHERE id = ?
+				if match[1] == "" {
+					parameters = append(parameters, 0)
+				}
+
 				position, _ := strconv.Atoi(match[1]) //nolint:errcheck
 				if position > len(parameters) {
 					parameters = parameters[:position]

--- a/options_test.go
+++ b/options_test.go
@@ -6,6 +6,7 @@ import (
 	"testing"
 
 	"github.com/lib/pq/oid"
+	"github.com/stretchr/testify/assert"
 )
 
 func TestInvalidOptions(t *testing.T) {
@@ -27,6 +28,39 @@ func TestInvalidOptions(t *testing.T) {
 			}
 
 			t.Error("unexpected pass")
+		})
+	}
+}
+
+func TestSimpleQueryParameters(t *testing.T) {
+	type test struct {
+		query      string
+		parameters []oid.Oid
+	}
+
+	tests := map[string]test{
+		"positional": {
+			query:      "SELECT * FROM users WHERE id = $1 AND age > $2",
+			parameters: []oid.Oid{0, 0},
+		},
+		"unpositional": {
+			query:      "SELECT * FROM users WHERE id = ? AND age > ?",
+			parameters: []oid.Oid{0, 0},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			option := SimpleQuery(nil)
+
+			srv := &Server{}
+			err := option(srv)
+			assert.NoError(t, err)
+
+			statement, parameters, err := srv.Parse(context.Background(), test.query)
+			assert.NoError(t, err)
+			assert.NotNil(t, statement)
+			assert.Equal(t, test.parameters, parameters)
 		})
 	}
 }

